Migrate scores and beatmaps to realm

peppy opened the pull request:
This marks the final step of migrating osu! to realm. There's a bit to go through so let's get started.

# Scope of this PR

While I tried my best with prior PRs getting things in the correct places as much as possible, this PR is a bit of a mess of changes that could not be separated out without immense effort. As such, this is going to be a bit of a chore to review. Rather than leaving a length review dispersed throughout the files, I'd suggest initial reviews along the lines of:

- Testing things work by running through common scenarios and pointing out any high level issues
- Speed-reading through the code changes and bringing to light any overarching concerns

If you need any further details on why decisions were made or want to talk through things on voice chat, I'll be available for that too.

I've also tried to tidy things up as much as possible, but there's definitely going to be some stray logging/comments/what not.

# Major changes

## EF contexts are no longer created for fresh installs

For existing installs, they will continue to be created on startup to perform migrations until we are happy that all data has been migrated correctly. That can likely be done soon after this goes live and we have confirmed things are in a good state, and followed up with a deletion of the EF `client.db` file post-migration.

~~Note that this PR does *not yet migrate scores and beatmaps*. I will push those changes in the coming days (wanted to get this PR out for review as-is since most testing will be done in isolation, I imagine).~~ Has been added.

## EF classes have been renamed

In order to make the fewest number of changes in this PR, realm models have been dropped in place of EF ones, with the EF classes remaining around (for migration purposes - they can eventually be removed too). As such, naming is not final and I'd ask that you don't pick apart the naming as this can be done in follow-up work.

## `Detach` and AutoMapper are used extensively

In order to get things working with minimal code changes, towards the end of this PR I turned to automapper to handle the remaining flows. Searching for usages of `.Detach` is recommended to see the locations where this is required. Note that some of these usages are redundant (detach is a noop if the target is already non-managed) but added for safety.

The eventual goal will be to remove many of these usages (not yet sure if all can be removed yet). This will require some restructuring of models (like `PlaylistItem`) and components (like `BeatmapCarousel`). I already have a pretty good idea of how things should be done in `BeatmapCarousel`, so that will likely be my next focus. It's made tricky by things like difficulty calculation, with calls nested deep within that hierarchy at the moment.

Note that in this PR, *all* query methods still remaining on `Manager` classes either return detached or `ILive` instances.

I have only tested so far using a modest collection of 190 beatmap sets. Performance testing (loading song select, and startup overhead of `MusicController`) is encouraged. I'm not too concerned as long as it is within reasonable error margins of EF - ie. we want to be able to deploy this without users having >10 second load times. Performance can be vastly improved in follow-up work.

## Editor save support is done via AutoMapper copying

Because most usages of `BeatmapInfo` are detached, the easiest way to get the existing save flows working was to use AutoMapper copy updated pieces via a "map" operation. Because this is very localised to the beatmap saving flow, I've only added support for it for beatmaps (see `RealmObjectExtensions.CopyChangesToRealm`). I don't foresee this existing far into the future.

Replacing this should not be high effort and I already have a good idea of the direction we should be taking to do so.


## Realm subscriptions are used in place of event flow for hiding/deletion

This logic has been added in several places. One important thing to note is that due to realm subscriptions not providing a frozen snapshot, it is impossible to look up "deleted" indices, even in the case that they are still in the realm. An example would be our `DeletePending` flag, where we'd expect to be able to still lookup pending deletions when the disappear from a subscription with a `s => !s.DeletePending` filter applied, but can't.

This looks to be something which comes up quite often in realm discussions/issues, and [they have shown interest in solving it](https://github.com/realm/realm-dotnet/issues/2752#issuecomment-1004891693) quite recently so there is hope.

To work around this, in some cases (such as `BeatmapCarousel`) I have opened a second subscription to track the inverse condition and then used the `InsertedIndices` to track the "deleted" or "hidden" items.

## `RealmContextFactory` is used directly in many places `BeatmapManager` or `ScoreManager` was previously used

I think going forward we will be wanting to do this in more places. It's the direction I took with already-merged realm work so this should come as no surprise. I'm mentioning it here because in several test scenes that use local instances of `BeatmapManager`, it is an added requirement to re-cache the `RealmContextFactory` as well to ensure correct behaviour.

## I'm not sure about `RealmUser`

The user class is in a bit of a weird place, and I'm not sure if/how it should exist. For instance, it's current an `EmbeddedObject` which means the same user may exist multiple times across different beatmaps and scores. Probably not what we want.

I didn't address this any further in this PR as user handling was one of the pain points of my work (due to it being used in `ScoreInfo` in various ways). We can change this going forward as required - migration should be quite easy.

This also means that `RealmUser` is used in a lot of places we don't ever expect it to persist (in place of `APIUser`).

## `ScoreInfo.Statistics` handling is ugly

Yeah. Work is still required around `ScoreInfo` as it is still being used to ferry multiple different usages of scores. Open to suggestion, but probably best left for follow-up work too.

## Usage of [realm query language](https://docs.mongodb.com/realm-legacy/docs/javascript/latest/api/tutorial-query-language.html)

Many usages of `IQueryable` have been converted to using the realm query language. This is usually due to something not being supported otherwise (nested lookups is the most common one). 

We're probably going to see more of this going forward, for instance as we add more advanced queries in the beatmap carousel to optimise filtering. It doesn't seem too bad.

## Nullability of `BeatmapInfo.BeatmapSet`

Not sure about this one. We may want to make it non-null. I believe there's a least one other case of unsure nullability. If we choose to make it non-null, we will need to bypass it during the construction of new beatmaps, or create a placeholder for such cases.

## Scope of testing

I have tested quite comprehensively with local gameplay and multiplayer. I haven't tested solo score submission and probably a heap of edge cases. Appreciate further testing.

Note that you will need to delete your realm database when leaving this branch. Make sure to make a backup if you care.

Closes #7057.

https://user-images.githubusercontent.com/191335/149159336-69a94172-cd97-40e7-ac40-732bf41a08bb.mp4

peppy committed:
"Update" `BeatmapCarousel`

peppy committed:
"Update" `MusicController`

peppy committed:
Add basic `RealmScore` implementation

peppy committed:
Add score importer

peppy committed:
Rename EF classes to allow for shit to hit the fan

peppy committed:
Rename realm to become imposter classes

peppy committed:
Add required properties to make realm models backwards compatible

peppy committed:
Fix issues with editor check tests

peppy committed:
Update mock model usage to set `GUID`s instead of `int`s

peppy committed:
Update sets of `BeatmapSet.Metadata` to instead create a `Beatmap`

peppy committed:
Update null fallback cases involving `OnlineID`

peppy committed:
Update usages of `APIUser` to `RealmUser`

peppy committed:
Remove EF specific tests that have since been replaced

peppy committed:
Miscellaneous fixes that don't fit elsewhere

peppy committed:
Remove unnecessary assigns of `BeatmapSetInfo.Metadata`

peppy committed:
Remove unnecessary assigns of EF foreign `ID` fields in tests

peppy committed:
Fix incorrect `Metadata`-related null checks

peppy committed:
Mark cases where `BeatmapSet` is generally guaranteed to be non-null

peppy committed:
Replace EF `RulesetStore` with realm version

Pass full EF context factory to `RealmContextFactory` for migration purposes

peppy committed:
Use `RealmContextFactory` instead of EF

peppy committed:
Various updates to ruleset and primary key usages to move closer to realm support

peppy committed:
Remove EF `FileStore`

peppy committed:
Remove unused store classes

peppy committed:
Switch `BeatmapModelManager` to use `RealmArchiveModelManager` base class

peppy committed:
Update usages to consume `IRulesetStore`

peppy committed:
Replace EF `ScoreInfo` with realm version

May contain errors.

peppy committed:
Add statistics storage to realm model

peppy committed:
Add required properties for compatibility with existing code

peppy committed:
Update many score-related classes to move closer to being able to persist to realm

peppy committed:
Remove `BeatmapStore` and update surrounding code

peppy committed:
Update some more incorrect types for primary key access/set

peppy committed:
Remove unused `ScoreImporter` class

peppy committed:
more fixes (almost compiles, except ruleset and manager)

peppy committed:
Reimplement all query methods

peppy committed:
Remove old EF classes

peppy committed:
Update model manager and many related classes to get things compiling again

peppy committed:
Fix missing interface implementation of `IRulesetStore`

peppy committed:
Don't persist `Countdown` to realm for now

It's another enum which is a pain to handle, and not actually being
consumed anywhere.

peppy committed:
Fix many shortcomings and compatibility issues with EF classes post-rename

peppy committed:
Add back missing `IRulesetStore` cache

peppy committed:
Apply NRT to `BeatmapManager` and move `Hide`/`Restore` methods across

peppy committed:
Fix intro screen

Fix things

peppy committed:
Start introducing `ILive`

peppy committed:
Inline query methods from `BeatmapModelManager` to `BeatmapManager` where possible

peppy committed:
Remove unnecessary re-query of beatmap set in editor menu construction

peppy committed:
Remove unused `includeProtected` parameter

peppy committed:
Make tests compile again

peppy committed:
Replace `BeatmapLeaderboard` event flow with realm subscriptions

peppy committed:
Update `ToLive` usages in line with recent changes

peppy committed:
Replace `BeatmapDownloadTracker` event flow with realm subscriptions

peppy committed:
Replace `ScoreDownloadTracker` event flow with realm subscriptions

peppy committed:
Fix missing disposal of realm subscriptions in `BeatmapCarousel`

peppy committed:
Remove updated/removed flow method mapping

peppy committed:
Replace `TopLocalRank` event flow with realm subscriptions

peppy committed:
Replace `OnlinePlayBeatmapAvailabilityTracker` event flow with realm subscriptions

peppy committed:
Replace `SpectatorScreen` event flow with realm subscriptions

peppy committed:
Fix invalid equality comparison in `BeatmapLeaderboard`

peppy committed:
Fix missing `BeatmapMetadata.ToString`

This is relied on by a few usages.

peppy committed:
Fix invalid DI resolution of `RealmFileStore`

peppy committed:
Fix `BeatmapInfo` file lookup not handling the case where no files exist

Quite common for test scenes.

peppy committed:
Add back settable `RulesetID` for now

peppy committed:
Fix `Max` lookup methods not checking for zero beatmap count

peppy committed:
Temporarily disable `WorkingBeatmapCache` and fix multiple invalid data flows

peppy committed:
Fix skin lookup when there's no beatmap file available

peppy committed:
Fix some null inspections

peppy committed:
Fix some remaining test failures

peppy committed:
`BeatmapSetInfo` detach support

peppy committed:
Update realm queries to use `Filter` to allow for indirect property filtering

peppy committed:
"Fix" OnlinePlayBeatmapAvailabilityTracker

peppy committed:
Fix rulesets being out of order

peppy committed:
Fix potential nullref when song select filters to no results

peppy committed:
Detach and reattach scores to make work

peppy committed:
Handle ignored mappings locally in `Detach` configuration

peppy committed:
Fix incorrect ordering and grouping of difficulties at song select

peppy committed:
Fix results screen test failures due to relation query

peppy committed:
Fix `BeatmapDifficultyCache` not working with detached beatmaps

peppy committed:
Better handle Statistics to avoid losing data

peppy committed:
Fix multiple tests via null checks and changing `ToLive` to `Detach`
flow

peppy committed:
Reattach detached items on delete/undelete

peppy committed:
Fix OnlineAvailabilityTracker referencing a value in query that could potentially be null

peppy committed:
Remove incorrect test fixture specification

peppy committed:
Update `FilterMatchingTest` and filter code to use ruleset's `OnlineID`

peppy committed:
Fix incorrect `BeatmapInfo.ToString` implementation

peppy committed:
Fix incorrect access to `ILive` in `BeatmapSkinResources` tests

peppy committed:
Add missing `Ruleset` in `ReplayRecorder` tests

peppy committed:
Fix `TestScenePresentScore` sharing metadata/difficulty across multiple beatmaps

peppy committed:
Improve legibility of `RulesetInfo.Equals`

peppy committed:
Ensure `Score` created by `GameplayState` has a valid ruleset

peppy committed:
Fix `BeatmapInfo.Clone` potentially not cloning if already detached

peppy committed:
Mention safety failures of Beatmap/Score constructors

peppy committed:
Remove assertion of `ScoreInfo.Combo` being database persisted

peppy committed:
Fix many instances of `User` being null in score import tests

peppy committed:
Detach score during import tests to ensure original object doesn't get managed

peppy committed:
Fix `ScoreInfo` oversights causing automapper to fail

Parameter in ctor *has* to be named `realmUser` else automapper will try
to map to the `User` property.

peppy committed:
Fix unsupported realm operations in multiple tests

peppy committed:
Update handling of ruleset nullability when handling a game-wide change

peppy committed:
Reimplmeent `IsAvailableLocally` as an `abstract` method

peppy committed:
Remove test coverage of scores being deleted when beatmaps are

This is not supported in realm for now. Probably best suited to a
separate pass, similar to files, using backlink count.

peppy committed:
Add support for persisting score's mods to realm

peppy committed:
Fix incorrect conditional for deciding whether scores can be deleted from UI

peppy committed:
Fix multiple tests with incorrect access to beatmap imports

peppy committed:
Add note about `BeatmapMetadata.Author` being weird

peppy committed:
Fix beatmap carousel test failures

peppy committed:
Remove unnecessary DI dependencies from cache test

peppy committed:
Fix protected beatmaps showing up in the song select carousel

peppy committed:
Add missing "non-null" elements missing from some tests

peppy committed:
Bind ruleset to toolbar later for safety

peppy committed:
Add `RulesetInfo` hashcode implementation and tidy up equality

peppy committed:
Fix a couple of minor issues with `TestSceneBeatmapRecommendations`

peppy committed:
Detach rather than consume live when presenting a beatmap

peppy committed:
Fix donwload trackers not considering deleted scores

peppy committed:
Fix `OnlinePlayBeatmapAvailabilityTracker` not correctly tracking beatmap import changes

peppy committed:
Fix more incorrect test access to `ILive`

peppy committed:
Use better method to ensure online availability tracker is in a clean state

peppy committed:
Fix edge cases in online availability tracker and combine query code

peppy committed:
Ensure `BeatmapInfo` `Difficulty` and `Metadata` is non-null

peppy committed:
Update null allowances across beatmaps and scores

peppy committed:
Fix one missed instance of `GetResultSafely`

peppy committed:
This file should have been deleted in a previous commit (rebase failure)

peppy committed:
Fix `BeatmapDifficulty` cloning regression

peppy committed:
Add back caching in `WorkingBeatmapCache`

peppy committed:
Fix incorrect realm factory isolation in `TestScenePlaySongSelect`

peppy committed:
Fix `TestScenePlaySongSelect` failure due to detach clone depth

peppy committed:
Fix another case of test ruleset without overriding `ShortName` primary key

peppy committed:
Fix custom import process in `TestSceneDrawableRoomPlaylist` not working with realm

peppy committed:
Allow `RealmArchiveModelManager` file operations to be performed on detached instances

peppy committed:
Fix incorrect realm access after new beatmap import

peppy committed:
Remove incorrect test assertion

peppy committed:
Update `BeatmapModelManager.Save` to work for editor scenarios

peppy committed:
Make `RealmArchiveModelManager.Update` work using automapper

peppy committed:
Add basic write support via automapper

peppy committed:
Add protections against test null refs when beatmap load fails

peppy committed:
Fix most remaining test issues

peppy committed:
Fix intro screen hitting null reference if intro beatmap is unavailable

peppy committed:
Fix multiple remaining warnings

peppy committed:
Remove hide/restore event flow

peppy committed:
Clean up `BeatmapManager` query methods

peppy committed:
Update migration test to use realm file as test

peppy committed:
Localise EF context factory usage to migration only

peppy committed:
Add refetch for non-managed hide/restore attempts

peppy committed:
Fix beatmap mass deletion flow

peppy committed:
Fix beatmap restore/undelete flows

peppy committed:
Fix deleted/hidden carousel queries

peppy committed:
Fix protected beatmap sets getting deleted

peppy committed:
Update `MusicController` to handle deletions more correctly

peppy committed:
Fix song select import popup not always showing

frenzibyte committed:
Handle automapper realm cyclic references via `AfterMap`s

This may not be the cleanest solution, but there don't seem to be any
way towards this either.

- `UseDestinationValue` has been inherited by default as noted in
https://docs.automapper.org/en/stable/10.0-Upgrade-Guide.html#usedestinationvalue-is-now-inherited-by-default, and its behaviour in this case would be using the nested **managed** realm object for the destination member rather than creating an unmanaged version.

- `MaxDepth` already sets `PreserveReferences` so there's no point of using it.

- `MaxDepth` should probably not be set for all maps, only for those with
cyclic references, to avoid the expensive overhead of `PreserveReferences`, as mentioned in https://docs.automapper.org/en/stable/5.0-Upgrade-Guide.html#circular-references.

That aside, `MaxDepth` should actually only be set to `1` for
`BeatmapSetInfo` mapping, because we don't want AutoMapper to create a
nested instance of `BeatmapSetInfo` in each mapped/detached beatmap, but
for some reason, doing that will cause automapper to not map any beatmap
inside the set and leave it with 0 beatmaps.

While on the other hand, using `MaxDepth(2)` for `BeatmapSetInfo` works,
but creates an unused instance of a `BeatmapSetInfo` inside each mapped
beatmap, which may not be ideal.

For `BeatmapInfo`, it has to be `MaxDepth(2)`, in which the first
`BeatmapInfo` depth would be itself (when detaching a beatmap), and the
second would be nested beatmaps inside the mapped/detached
`BeatmapSetInfo` within the beatmap. (note that when detaching a beatmap
set, the unused instance of `BeatmapSetInfo` within each beatmap of that
beatmap set doesn't also have a list of unused beatmaps as one might expect from the depth specification, it surprisingly has 0 beatmaps)

This causes it to create an unused instance of `BeatmapInfo` in the beatmap set resembling the root mapped/detached beatmap, but that one might be inevitable.

peppy committed:
Add equatable support to `IUser` and `RealmUser`

Not sure this will stick, but let's add it for now to make testing
detach support work nicely.

peppy committed:
Update detach test to assert correct behaviour

peppy committed:
Add test of cyclic beatmap/beatmapset references

peppy committed:
Combine mapper configurations and add more explanation about special cases

peppy committed:
Add test coverage of updating via copying changes from detached instance

peppy committed:
Move manager `Update` methods to be explicit to where they are still used by legacy code

Also fixes skin hash repopulation being completely broken.

peppy committed:
Make copying detached changes to realm only exposed for `BeatmapSet`

Also fixes remaining issues with the copy process.

peppy committed:
Ensure scores are cleaned up alongside beatmap so they don't have a null reference

peppy committed:
Sanitise and remove some usages of `Detach` which are no longer required

peppy committed:
Cascade delete metadata when beatmaps are deleted

peppy committed:
Merge branch 'master' into realm-integration/score-and-beatmaps

peppy commented:
Oh and there's no doubt going to be some (null related) warnings that show up (I have my rider set to non-intensive mode). Will fix those in the coming hours/days.

peppy committed:
Replace compatibility properties with direct references

peppy committed:
Remove stay newline

peppy committed:
Add missing ruleset in test scores

peppy committed:
Merge branch 'master' into realm-integration/score-and-beatmaps

peppy committed:
Fix CI reported warnings

peppy committed:
Add far too many fixes for ruleset non-nullable requirements

peppy committed:
Fix further warnings

peppy committed:
Fix potential mod nullref in `APIUserScoreAggregate`'s `CreateScoreInfo` implementation

bdach commented on osu.Game/Scoring/ScoreInfo.cs:
```diff
@@ -3,191 +3,235 @@
-                    return localAPIMods;
+                if (localAPIMods == null)
+                {
+                    // prioritise reading from realm backing
+                    if (!string.IsNullOrEmpty(ModsJson))
+                        localAPIMods = JsonConvert.DeserializeObject<APIMod[]>(ModsJson);
 
-                if (mods == null)
-                    return Array.Empty<APIMod>();
+                    // then check mods set via Mods property.
+                    if (mods != null)
+                        localAPIMods = mods.Select(m => new APIMod(m)).ToArray();
+                }
```
You've mentioned `Statistics` being quite ugly in the OP but this also needs to go too eventually. It's the same sort of badness.

peppy replied:
For sure.

I've done a basic clean-up pass on these to bring a bit more sanity to the variable naming (and make them both match as close as possible to each other in structure).

bdach commented on osu.Game/Skinning/SkinModelManager.cs:
```diff
@@ -262,5 +262,7 @@ public void Save(Skin skin)
             });
         }
+
+        public override bool IsAvailableLocally(SkinInfo model) => false;
```
bit of a weird default to leave without a TODO at least

peppy replied:
Think it would be better to default to `true`? I've changed to that and added a comment.

bdach replied:
`true` is better for sure as far as I'm concerned.

bdach commented on osu.sln.DotSettings:
```diff
@@ -789,7 +789,7 @@ See the LICENCE file in the repository root for full licence text.&#xD;
 	<s:String x:Key="/Default/CodeStyle/Naming/XamlNaming/UserRules/=XAML_005FRESOURCE/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
 	<s:String x:Key="/Default/CustomTools/CustomToolsData/@EntryValue"></s:String>
-	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EFeature_002EServices_002EDaemon_002ESettings_002EMigration_002ESwaWarningsModeSettingsMigrate/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EFBeature_002EServices_002EDaemon_002ESettings_002EMigration_002ESwaWarningsModeSettingsMigrate/@EntryIndexedValue">True</s:Boolean>
```
looks like a stray rename from an `EF` replace

peppy replied:
Oops, this was on my TODO list to fix and I forgot about it.

bdach reviewed (commented):
here's an initial review pass. tried to keep to the important stuff. there were a few more nits but i deleted those comments for now because i don't know how else to handle this diff. and gonna be honest, the quality of this review is probably questionable anyway because i'm barely awake. looked generally inoffensive though.

haven't ran this at all in any capacity yet.

bdach commented:
as for the opening post - agree that `RealmUser` and `ScoreInfo` need general work but like... nothing can conceivably happen in those regards in this pull. so i didn't really even attempt to question those for now.

i'm treating this pull as something that should result in a state that's no worse than what's already there and from then on the rest can hopefully be improved.

peppy commented on osu.Game/Beatmaps/BeatmapDifficulty.cs:
```diff
@@ -32,25 +34,27 @@ public BeatmapDifficulty(IBeatmapDifficultyInfo source)
         public BeatmapDifficulty Clone()
         {
-            var diff = (BeatmapDifficulty)Activator.CreateInstance(GetType());
+            var diff = (BeatmapDifficulty)MemberwiseClone();
```
I did originally have `Detach` calls in these lines as a safety measure. Not sure, will revisit this and see if we can still use `Activator` or just `new` instead.

peppy replied:
I guess we were using `CreateInstance` for the one taiko exception where this class is derived?

How would you feel about making this method `virtual` with the base implementation using `new BeatmapDifficulty()`?

bdach replied:
Yeah, that newest version looks fine.

peppy commented on osu.Game/Database/RealmObjectExtensions.cs:
```diff
@@ -51,21 +128,34 @@ public static class RealmObjectExtensions
         }
 
-        public static ILive<T> ToLiveUnmanaged<T>(this T realmObject)
+        public static ILive<T> ToLiveUnmanaged<T>(this T realmObject
+        )
```
I also recall there being no usages of this, so I may look to remove it instead.

peppy replied:
Never mind, that was the `IEnumerable` version. Will just leave it in place for now.

peppy committed:
Fix unintended change to dotsetting

peppy committed:
Change default value and add comment explaining why skins are never "locally available"

peppy committed:
Sanitise mods / statistics cache logic in `ScoreInfo`

peppy committed:
Fix stray bracket

peppy commented on osu.Game/OsuGameBase.cs:
```diff
@@ -460,37 +445,16 @@ private void onRulesetChanged(ValueChangedEvent<RulesetInfo> r)
             AvailableMods.Value = dict;
         }
 
-        private void runMigrations()
```
For sure, I'll add that to a post-merge TODO (which we'd probably want added before pushing a release with these changes). Can agree with renaming the realm before creating a new one, and showing a notification to the user to let them act on it further.

peppy committed:
Add back `PerformRead` return safety by checking `IsManaged` status of returned data

peppy committed:
Remove unnecessary `Ruleset` null check in `BeatmapDifficultyCache`

peppy commented on osu.Game/Beatmaps/BeatmapDifficultyCache.cs:
```diff
@@ -136,7 +136,7 @@ protected override void LoadComplete()
 
             // Difficulty can only be computed if the beatmap and ruleset are locally available.
-            if (localBeatmapInfo?.IsManaged != true || localRulesetInfo == null)
+            if (localBeatmapInfo?.Ruleset == null || localRulesetInfo == null)
```
There was a point in time where the `Ruleset` was null in many cases so it needed to be checked here. I think that has been resolved in the last few days of work so this is no longer required (and yes, this becomes a `localBeatmapInfo==null` check). Will make that change.

peppy commented on osu.Game/Database/RealmLive.cs:
```diff
@@ -61,8 +61,9 @@ public void PerformRead(Action<T> perform)
         {
-            if (typeof(RealmObjectBase).IsAssignableFrom(typeof(TReturn)))
-                throw new InvalidOperationException(@$"Realm live objects should not exit the scope of {nameof(PerformRead)}.");
+            // TODO: this is weird and kinda wrong... unmanaged objects should be allowed?
```
Good call, I've made this change.

peppy committed:
Use `ctor` rather than `MemberwiseClone` to guarantee a safer clone of `BeatmapDifficulty`

frenzibyte commented on osu.Game/Beatmaps/BeatmapInfo.cs:
```diff
@@ -2,180 +2,195 @@
-        [JsonProperty("difficulty_rating")]
-        [Column("StarDifficulty")]
-        public double StarRating { get; set; }
+        [Ignored]
+        public CountdownType Countdown { get; set; } = CountdownType.Normal;
```
Was it intentional leaving this as-is? sounds like it could be worked around in a similar fashion to `Status`/`StatusInt`.

peppy replied:
Yeah, explained in https://github.com/ppy/osu/pull/16428/commits/d7fe3584cde52011d776d26f1bcb95a651d56896.

peppy committed:
Bump realm schema version to allow upgrades

peppy committed:
Fix deleted scores not being cleaned up on next startup

frenzibyte reviewed (dismissed):
I've started to look into this from the high-level testing side, and I've caught a few issues:
 1. Initial load of song select displays beatmaps of all rulesets even with `Show converted` turned off, triggering one criteria filter fixes things back (like in the video by changing global ruleset).    

    https://user-images.githubusercontent.com/22781491/149266604-e9e5774e-84b0-47f2-9e51-3af2d96f2279.mp4

 2. Beatmaps don't get populated with the online status from `BeatmapOnlineLookupQueue` (in fact, it doesn't seem like `BeatmapOnlineLookupQueue` is used at all in this branch).
 3. Scores do not get purged from the database after removing it in-game, missing cleanup logic.
 4. Exported scores get named `ScoreInfo (unmanaged).osr` rather than the score's actual name. (tested by playing a beatmap, then exporting the resultant score)
 5. After importing an exported score, viewing the imported score's panel by clicking on the "finished importing" notification displays the replay download button in a "download" state rather than "locally available" state.

peppy committed:
Fix typo in comment in `IntroScreen`

peppy commented:
Thanks for the testing. The online lookup queue not being fired is also a reason why imports feel faster on this branch (and also why older beatmaps don't ever finish importing - because they don't get their online ID at the end of the process).

peppy committed:
Fix filter criteria not being applied after carousel loads new beatmap sets

peppy committed:
Remove oudated comments

peppy committed:
Add synchronous fetch flow to `BeatmapOnlineLookupQueue`

The async flow doesn't work great with the realm import process. We
might be able to improve on this going forward, but for the time being
adding a synchronous path seems safest.

After all, we are already an an asynchronous (dedicated) thread pool at
this point.

peppy committed:
Remove pointless override method in `RealmArchiveModelManager`

peppy committed:
Remove `async` from `Populate` method

peppy committed:
Fix scores not using correct filename/display strings

I've updated all cases where we should have been using
`GetDisplayString()` anyway, but left the `ToString()` implementations
in place for safety. They should probably be removed in the future.

peppy commented:
> 1. Initial load of song select displays beatmaps of all rulesets even with `Show converted` turned off, triggering one criteria filter fixes things back (like in the video by changing global ruleset).

Fixed in 6025fe325df0d49e5526b697d514d86c0788259a

> 2. Beatmaps don't get populated with the online status from `BeatmapOnlineLookupQueue` (in fact, it doesn't seem like `BeatmapOnlineLookupQueue` is used at all in this branch).

Fixed in 93c78253d65065ba649f382183d4e12b643ab44a (see the two prior commits for more related changes to avoid realm access across multiple threads)

> 3. Scores do not get purged from the database after removing it in-game, missing cleanup logic.

Fixed via dcc354aa7c24640c8aceef89623c0efce472b273

> 4. Exported scores get named `ScoreInfo (unmanaged).osr` rather than the score's actual name. (tested by playing a beatmap, then exporting the resultant score)

Fixed via c61419dfe5606b2749381b9a7eae275d4762e4bb

> 5. After importing an exported score, viewing the imported score's panel by clicking on the "finished importing" notification displays the replay download button in a "download" state rather than "locally available" state.

I couldn't repro this one. May have been fixed by one of the changes along the way. Can you give it another test?

frenzibyte commented:
Can still reproduce on latest changes (note that the exported score was from a lazer score that I made mysself prior to recording this):

https://user-images.githubusercontent.com/22781491/149294314-5e3022f3-6a6b-4141-9d03-43f5e5a83fe7.mp4

peppy committed:
Add EF to realm score migration

peppy committed:
Remove pointless compatibility parameter `BeatmapSetInfoID`

peppy committed:
Add EF to realm beatmap migration

peppy committed:
Move `RulesetStore` construction earlier in process so rulesets are available for EF migration

peppy committed:
Merge branch 'master' into realm-integration/score-and-beatmaps

peppy committed:
Uncomment EF deletion lines in migrations

peppy commented:
I've also added EF migrations for beatmaps and scores.

peppy commented:
@frenzibyte Looks something to do with the filename not being `replay.osr`, on a quick guess.

peppy committed:
Remove one more unnecessary `Detach` operation

peppy committed:
Fix delete/clear scores buttons not working

peppy commented:
I still can't repro the score failure even when setting a new score. Can you take a look into it further from your end? If you didn't set the score a second time it may have imported in a bad state, not sure.

bdach commented on osu.Game/Beatmaps/BeatmapOnlineLookupQueue.cs:
```diff
@@ -52,6 +54,12 @@ public BeatmapOnlineLookupQueue(IAPIProvider api, Storage storage)
                 prepareLocalCache();
         }
 
+        public void Update(BeatmapSetInfo beatmapSet, CancellationToken cancellationToken)
```
`cancellationToken` is not at all used here...?

bdach commented on osu.Game/Beatmaps/BeatmapOnlineLookupQueue.cs:
```diff
@@ -71,40 +79,38 @@ private void lookup(BeatmapSetInfo set, BeatmapInfo beatmapInfo)
                 // intentionally blocking to limit web request concurrency
                 api.Perform(req);
 
+                if (req.CompletionState == APIRequestCompletionState.Failed)
+                {
+                    logForModel(set, $"Online retrieval failed for {beatmapInfo}");
+                    beatmapInfo.OnlineID = -1;
+                }
```
This looks as if there's a missing early return? Is there a need to check `Response` at all if the state is failed?

peppy replied:
Arguably the response would always be null in this case, but I've changed it to return for safety.

bdach reviewed (commented):
non-minor things from a skim of the latest changes.

still haven't tested yet. probably going to happen during the weekend if this isn't merged by then.

peppy committed:
Ensure detached when performing model `Clone` operations on `BeatmapInfo`/`ScoreInfo`

peppy commented on osu.Game/Beatmaps/BeatmapInfo.cs:
```diff
@@ -2,180 +2,192 @@
+
+        public int BeatmapVersion;
+
+        public BeatmapInfo Clone() => (BeatmapInfo)MemberwiseClone();
```
I've added back `Detach` in both these cases for safety. The only reason I originally removed it was to reduce "unnecessary" `Detach` calls in the project.

peppy committed:
Remove unused `cancellationToken` parameter in synchronous `BeatmapOnlineLookupQueue` flow

peppy committed:
Return immediately on failed web request in synchronous `BeatmapOnlineLookupQueue` flow

peppy committed:
Fix watch replay button not working immediately after playing

smoogipoo commented on osu.Game/Tests/Visual/RateAdjustedBeatmapTestScene.cs:
```diff
@@ -12,8 +12,11 @@ protected override void Update()
 
-            // note that this will override any mod rate application
-            Beatmap.Value.Track.Tempo.Value = Clock.Rate;
+            if (Beatmap.Value.TrackLoaded && Beatmap.Value.Track != null) // null check... wasn't required until now?
```
Forgotten about comment? Not sure about this - you'd would hope track is non-null (at the very minimum it should be a virtual track), so something looks wrong.

peppy replied:
Looks to have since been resolved, have reverted.

peppy committed:
Only copy across `Hash` and `ID` so statistics aren't lost

smoogipoo commented on osu.Game/Database/EFToRealmMigrator.cs:
```diff
@@ -30,6 +34,171 @@ public void Run()
+                                Bookmarks = beatmap.Bookmarks,
+                                Ruleset = realm.Find<RulesetInfo>(beatmap.RulesetInfo.ShortName),
+                                Difficulty = new BeatmapDifficulty(beatmap.BaseDifficulty),
+                                Metadata = new BeatmapMetadata
+                                {
+                                    Title = beatmap.Metadata.Title,
+                                    TitleUnicode = beatmap.Metadata.TitleUnicode,
```
This crashes at startup for me because it's importing protected intro beatmaps which have a null metadata:
![image](https://user-images.githubusercontent.com/1329837/149493767-7b650005-7133-44f4-9b46-04af88859d73.png)
Or something else is going on, I don't know, because I have a _lot_ of beatmap sets where at least one contained beatmap has null metadata:
![image](https://user-images.githubusercontent.com/1329837/149494478-916c423c-5e58-47e4-a7d5-e36527bcf1b0.png)
Which is significantly more beatmap sets than those where all beatmaps have metadata:
![image](https://user-images.githubusercontent.com/1329837/149494595-e263cce8-6a72-4fde-b123-2c7127e44588.png)

[client.db.zip](https://github.com/ppy/osu/files/7869436/client.db.zip)

peppy replied:
we'll need to account for that. set level was left null during import when it didn't match across all beatmaps.

i have a separate branch that does similar deduping in realm, but that's for another day.

peppy commented on osu.Game/Screens/Menu/IntroScreen.cs:
```diff
@@ -80,27 +81,36 @@ protected IntroScreen([CanBeNull] Func<MainMenu> createNextScreen = null)
                 {
                     setInfo = beatmaps.QueryBeatmapSet(s => s.ID == sets[RNG.Next(0, sets.Count - 1)].ID);
-                    initialBeatmap = beatmaps.GetWorkingBeatmap(setInfo.Beatmaps[0]);
+                    setInfo?.PerformRead(s =>
+                    {
+                        if (s.Beatmaps.Count == 0)
+                            return;
+
+                        initialBeatmap = beatmaps.GetWorkingBeatmap(s.Beatmaps[0].ToLive(realmContextFactory));
+                    });
```
Yeah this looks like remnants of before detach was being used, I'll update.

peppy committed:
Remove unnecessary `Live` conversion in `IntroScreen` (handled by `GetWorkingBeatmap`)

peppy committed:
Remove left-over methods in `MusicController`

peppy committed:
Add local `ContextFactory` caching to all remaining test scenes that create local managers

peppy commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapLeaderboard.cs:
```diff
@@ -43,7 +45,7 @@ protected override IReadOnlyDependencyContainer CreateChildDependencies(IReadOnl
             dependencies.Cache(rulesetStore = new RulesetStore(ContextFactory));
             dependencies.Cache(beatmapManager = new BeatmapManager(LocalStorage, ContextFactory, rulesetStore, null, dependencies.Get<AudioManager>(), Resources, dependencies.Get<GameHost>(), Beatmap.Default));
-            dependencies.Cache(scoreManager = new ScoreManager(rulesetStore, () => beatmapManager, LocalStorage, ContextFactory, Scheduler));
+            dependencies.Cache(scoreManager = new ScoreManager(dependencies.Get<RulesetStore>(), () => beatmapManager, LocalStorage, dependencies.Get<RealmContextFactory>(), Scheduler));
```
No reason, was older code.

peppy committed:
Update one more mismatched test implementation

peppy committed:
Remove outdated comment

peppy commented on osu.Game.Tests/Visual/Multiplayer/TestSceneMultiplayerQueueList.cs:
```diff
@@ -39,7 +39,7 @@ public class TestSceneMultiplayerQueueList : MultiplayerTestScene
         {
             Dependencies.Cache(rulesets = new RulesetStore(ContextFactory));
-            Dependencies.Cache(beatmaps = new BeatmapManager(LocalStorage, ContextFactory, rulesets, null, audio, Resources, host, Beatmap.Default));
+            Dependencies.Cache(beatmaps = new BeatmapManager(LocalStorage, ContextFactory, rulesets, API, audio, Resources, host, Beatmap.Default));
```
I've updated every similar case for now.

peppy committed:
Remove unnecessary null check and associated comment

peppy committed:
Remove pointless `Metadata` set

peppy committed:
Fix importer not considering that some EF beatmaps have no local metadata

peppy commented:
> Changing gameplay keybinds doesn't seem to work in this PR (e.g. after changing osu! buttons I can still only tap with Z & X)

Good catch. Can replicate (related to `IsManaged` check).

peppy committed:
Fix custom keybindings not working due to incorrect use of `IsManaged` flag

peppy commented:
I've pushed a fix for the key binding issue. Can be separated into a prereq PR if preferred.

bdach committed:
Merge branch 'master' into realm-integration/score-and-beatmaps

GopoBoi commented:
Global leaderboards seem to be using scorev1 values for the scores (they match the website scores, changing score display mode doesn't change anything):
![LwGiLIZ](https://user-images.githubusercontent.com/69014805/149624477-0d554c4d-ac99-4b1f-85de-8c8efda2b9f2.png)

bdach commented:
There's a `ToString()` mishap with `RealmUser` on multiplayer room items:

![2022-01-15-151444_650x109_scrot](https://user-images.githubusercontent.com/20418176/149624826-87492d52-7b80-4b44-a7e4-c45a0d186514.png)
![2022-01-15-151452_1364x125_scrot](https://user-images.githubusercontent.com/20418176/149624828-7cdc9ffe-ae3d-4abf-abad-7df64cef7ac9.png)

(`metadataInfo.Author` at the end there is an `IUser`, should have a `.GetDisplayString()` call).

bdach committed:
Fix `GetDisplayTitleRomanisable()` relying on `ToString()` implementation

bdach commented:
also just for posterity: beware of running visual tests. they will also run the EF migration because that runs in `OsuGameBase`.

bdach commented on osu.Game/Scoring/ScoreManager.cs:
```diff
@@ -150,7 +158,7 @@ public async Task<long> GetTotalScoreAsync([NotNull] ScoreInfo score, ScoringMod
                 else
                 {
-                    if (score.BeatmapInfo.ID == 0 || difficulties == null)
+                    if (!score.BeatmapInfo.IsManaged || difficulties == null)
```
https://github.com/ppy/osu/pull/16428#issuecomment-1013687921 is caused by this change. This used to mean "this beatmap is available locally" which is not the same as `IsManaged` because of detaching. Requerying seems out of the question and I'm not sure where to take this. Seems that best that can be done is a `MaxCombo == 0` check after diffcalc is done but that feels dicey.

peppy replied:
Hmm, `MaxCombo` is actually nullable for some reason, so checking whether null or not should in theory be enough?

bdach replied:
It's not nullable on the `StarDifficulty` that diffcalc returns if I remember correctly, I was thinking to check that one.

peppy replied:
It looks like this conditional can be removed completely these days. I'll push this for some further testing, but it looks to work as expected locally in all the scenarios I can find.

bdach commented on osu.Game/Beatmaps/BeatmapDifficultyCache.cs:
```diff
@@ -136,7 +136,7 @@ protected override void LoadComplete()
 
             // Difficulty can only be computed if the beatmap and ruleset are locally available.
-            if (localBeatmapInfo?.IsManaged != true || localRulesetInfo == null)
+            if (localBeatmapInfo == null || localRulesetInfo == null)
```
This change on `localBeatmapInfo` is also not necessarily correct. For instance, if you open the [playlists results screen](https://github.com/ppy/osu/blob/4ca2822d409ff8b4b17213f9ef1c2916d8fdd706/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsResultsScreen.cs#L183) for a beatmap that is not happily available, a dummy beatmap info will happily make its way here and pass this check (which it would fail previously):

![2022-01-15-173420_1516x73_scrot](https://user-images.githubusercontent.com/20418176/149629762-57de61ba-e5e1-46a8-bef3-868c8a5f922f.png)

This causes the score panels on said playlists results screen to show 0 stars.

peppy replied:
Not 100% on this one. I don't think this is the underlying cause of that issue (see https://github.com/ppy/osu/issues/16480).

bdach commented:
Other issues I found in a run-through today:

- [ ] Opening the beatmap set overlay for seemingly any beatmap throws this:

```
2022-01-15 16:38:31 [error]: An unhandled error has occurred.
2022-01-15 16:38:31 [error]: System.NullReferenceException: Object reference not set to an instance of an object.
2022-01-15 16:38:31 [error]: at osu.Game.Overlays.BeatmapSet.Scores.TopScoreStatisticsSection.set_Score(ScoreInfo value) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Overlays/BeatmapSet/Scores/TopScoreStatisticsSection.cs:line 110
2022-01-15 16:38:31 [error]: at osu.Game.Overlays.BeatmapSet.Scores.DrawableTopScore..ctor(ScoreInfo score, Nullable`1 position) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Overlays/BeatmapSet/Scores/DrawableTopScore.cs:line 35
2022-01-15 16:38:31 [error]: at osu.Game.Overlays.BeatmapSet.Scores.ScoresContainer.<>c__DisplayClass27_2.<set_Scores>b__3() in /home/dachb/Dokumenty/opensource/osu/osu.Game/Overlays/BeatmapSet/Scores/ScoresContainer.cs:line 98
```

- [ ] Playlists leaderboards show 0 total score for every user:

![2022-01-15-173751_784x220_scrot](https://user-images.githubusercontent.com/20418176/149629966-20ac923f-8b5c-4ffb-8b37-fe1f23520868.png)

- [ ] Playlists results screen has score panels with 0* (mentioned above), but also no information about the beatmap (note missing artist/title/all other metadata):

![2022-01-15-173808_523x913_scrot](https://user-images.githubusercontent.com/20418176/149629977-2aafa976-61e2-4f26-9fe2-8804d11bafa2.png)

peppy committed:
Merge branch 'master' into realm-integration/score-and-beatmaps

peppy committed:
Fix null reference in equality comparison causing beatmap set overlay to crash

peppy committed:
Add back `null` beatmap allowance to `GetTotalScore` flow to fix playlist aggregate scores

peppy commented:
Not sure how to reply to the above comment so I'll just separate into one comment per issue

> Playlist results screen not showing beatmap info

This repros on master, it would seem.

https://github.com/ppy/osu/blob/2cd2b10ce1b1fd705a5d63fe007aba821745a5d4/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsResultsScreen.cs#L183

This looks to be the error site. It's using the global beatmap, which cannot be correct in the case of viewing results of a playlist item that isn't available locally.

peppy commented:
> Playlist leaderboards showing 0 score

This was due to the aggregate score having a `null` `BeatmapInfo`, against spec. Another case of usage of the no-parameter constructor which leads to stupid stuff happening. I've added back a null check in a3806f44a51adb35ad97299aa2250d1138b6ce18 to resolve this for now.

peppy commented:
> Beatmap set overlay crashes when showing a score

Fixed via 88bf406c225c3a074f2d9756b815338f42910bd9. Reference equality conversion to `.Equals` missing a null check (for the initial comparison - it is never null post that one).

peppy committed:
Remove incorrect `IsManaged` check in `ScoreManager`

peppy committed:
Remove one more usage of `IsManaged` which could potentially go wrong

peppy committed:
Initialise all parameters is paramaterless constructor for now for added safety

peppy committed:
Guard against null values getting inserted into database during score/beatmap imports

peppy committed:
Give the placeholder ruleset better defaults to allow tests to work again

peppy committed:
Update tests which were importing scores without a valid beatmap

peppy committed:
Fix incorrect realm (non-isolated instance) being used in two test scenes

peppy commented:
I've updated the way the parameterless constructors are written now to lead to more understandable results. Also added more safeties to ensure we don't write any nulls to the database on properties marked as non-null. Only null warning suppression should be on realm `IList`s now.

peppy commented:
I believe all reported issues have been fixed again.

bdach commented on osu.Game/Scoring/ScoreModelManager.cs:
```diff
@@ -63,6 +63,11 @@ protected override void Populate(ScoreInfo model, ArchiveReader? archive, Realm
             if (!model.Ruleset.IsManaged)
                 model.Ruleset = realm.Find<RulesetInfo>(model.Ruleset.ShortName);
 
+            // These properties are known to be non-null, but these final checks ensure a null hasn't come from somewhere (or the refetch has failed).
+            // Under no circumstance do we want these to be written to realm as null.
+            if (model.BeatmapInfo == null) throw new ArgumentNullException(nameof(model.BeatmapInfo));
+            if (model.Ruleset == null) throw new ArgumentNullException(nameof(model.Ruleset));
```
In light of 744084b4, will these ever realistically trigger if someone doesn't explicitly try to break the model by inserting a `null`?

peppy replied:
they shouldn't no

bdach commented:
On a half hour check of various things I couldn't find anything obviously broken that isn't already listed above as fixed (with the exception of the playlists total score thing which broke again).

peppy commented on osu.Game/Scoring/ScoreManager.cs:
```diff
@@ -132,9 +132,10 @@ public void GetTotalScore([NotNull] ScoreInfo score, [NotNull] Action<long> call
+            // TODO: This is required for playlist aggregate scores. They should likely not be getting here in the first place.
+            // ReSharper disable once ConditionIsAlwaysTrueOrFalse
+            if (score.BeatmapInfo == null)
+                return score.TotalScore;
```
Of course it has 😵‍💫

peppy replied:
I've pushed a fix for this for now. Was looking to add test coverage but ran into more (slightly unrelated) issues in the process, so that may come later.

peppy replied:
Test coverage added via #16489.

peppy committed:
Fix playlist aggregate scores not displaying (again)

peppy committed:
Merge branch 'master' into realm-integration/score-and-beatmaps

peppy committed:
Update automapper spec in line with v11

See https://docs.automapper.org/en/latest/11.0-Upgrade-Guide.html for
 more details.

peppy committed:
Fix test beatmaps created without hash being populated

peppy committed:
Improve reliability of mod deserialisation

peppy commented:
I've pulled in upstream changes, which includes a pretty substantial change to automapper (that I may have missed in the packages update branch). Switched away from using the `[IgnoreMap]` attribute to specifying as part of the mapper configuration.

Note that I did expand the scope of what is to be ignored as a result. In testing things look to still be working fine, but would recommend doing some brief re-testing at least.

peppy committed:
Remove one more outdated comment

peppy committed:
Fix `PLaylistsResultsScreen` tests falling over due to missing beatmap

smoogipoo committed:
Merge branch 'master' into realm-integration/score-and-beatmaps
